tower-http/src/cors.rs

```diff
@@ -0,0 +1,798 @@
+//!
+//! let mut service = ServiceBuilder::new()
+//!     .layer(cors)
+//!     .service_fn(handle);
+//!
+//! let request = Request::builder()
+//!     .header(header::ORIGIN, "https://example.com")
+//!     .body(Body::empty())
+//!     .unwrap();
+//!
+//! let response = service
+//!     .ready()
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(
+//!     response.headers().get(header::ACCESS_CONTROL_ALLOW_ORIGIN).unwrap(),
+//!     "*",
+//! );
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+
+use futures_core::ready;
+use http::{
+    header::{self, HeaderName, HeaderValue},
+    request::Parts,
+    Method, Request, Response, StatusCode,
+};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`Cors`] which adds headers for [CORS][mdn].
+///
+/// See the [module docs](crate::cors) for an example.
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+#[derive(Debug, Clone)]
+pub struct CorsLayer {
+    allow_credentials: Option<HeaderValue>,
+    allow_headers: Option<HeaderValue>,
+    allow_methods: Option<HeaderValue>,
+    allow_origin: Option<AnyOr<Origin>>,
+    expose_headers: Option<HeaderValue>,
+    max_age: Option<HeaderValue>,
+}
+
+const WILDCARD: &str = "*";
+
+impl CorsLayer {
+    /// Create a new `CorsLayer`.
+    ///
+    /// This creates a restrictive configuration. Use the builder methods to
+    /// customize the behavior.
+    pub fn new() -> Self {
+        Self {
+            allow_credentials: None,
+            allow_headers: None,
+            allow_methods: None,
+            allow_origin: None,
+            expose_headers: None,
+            max_age: None,
+        }
+    }
+
+    /// A very permissive configuration suitable for development:
+    ///
+    /// - Credentials allowed.
+    /// - All request headers allowed.
+    /// - All methods allowed.
+    /// - All origins allowed.
+    /// - All headers exposed.
+    /// - Max age set to 1 hour.
+    ///
+    /// Note this is not recommended for production use
+    pub fn permissive() -> Self {
+        Self::new()
+            .allow_credentials(true)
+            .allow_headers(Any)
+            .allow_methods(Any)
+            .allow_origin(Any)
+            .expose_headers(Any)
+            .max_age(Duration::from_secs(60 * 60))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Credentials`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    ///
+    /// let layer = CorsLayer::new().allow_credentials(true);
+    /// ```
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+    pub fn allow_credentials(mut self, allow_credentials: bool) -> Self {
```

Comment:
MDN recommends instead of setting it to false, just omit header.
I think better API would be `pub fn allow_credentials(mut self) -> Self` unconditionally setting it to `true`

Reply:
I generally find methods that take bools more flexible. Makes it easier to set it based on some other logic.

How about this e3c6ac81b2496 instead?

Reply:
Well it is up to you. MDN only _recommends_ to omit on `false` hence simplifying API is reasonable.
But making it conditional is ok too
It is just from user point of view `allow_credentials(false)` makes no sense because it is default behavior for when you do not send header 😄 
Hence flexibility actually is useless here

Reply:
Its mostly to allow things like `allow_credentials(std::env::var("SOME_ENV_VAR").is_ok())`. That also matches things like [`DefaultMakeSpan::include_headers`](https://docs.rs/tower-http/0.1.1/tower_http/trace/struct.DefaultMakeSpan.html#method.include_headers).

Reply:
Ah I see, makes sense

Reply:
> The only valid value for this header is true (case-sensitive). If you don't need credentials, omit this header entirely (rather than setting its value to false).

That also feels like stronger than just a recommendation 😅 Definitely something we should follow.

Reply:
Not really, RFC itself doesn't have strong words about it.

From https://datatracker.ietf.org/doc/html/rfc7480#section-5.6
```
By default, browsers do not send cookies when cross origin requests
   are allowed.  Setting the Access-Control-Allow-Credentials header
   field to "true" will send cookies.  Use of the
   Access-Control-Allow-Credentials header field is NOT RECOMMENDED.
```

@davidpdrsn note that it actually says nothing about `false` at all, and only recommends to not send it.

Reply:
Seems the current implementation is the way to go then.